runtime/src/legacymm/cpp/Memory.cpp

```diff
@@ -584,15 +583,78 @@ class ForeignRefManager {
   }
 };
 
+namespace {
+
+class ThreadLocalStorage {
+public:
+    using Key = void**;
+
+    void Init() noexcept { map_ = konanConstructInstance<Map>(); }
+
+    void Deinit() noexcept {
+        RuntimeAssert(map_->size() == 0, "Must be already cleared");
+        konanDestructInstance(map_);
+    }
+
+    void Add(Key key, int size) noexcept {
+        RuntimeAssert(storage_ == nullptr, "Storage must not be committed");
+        auto it = map_->find(key);
+        if (it != map_->end()) {
+            RuntimeAssert(it->second.second == size, "Attempt to add TLS record with the same key and different size");
+            return;
+        }
+        map_->emplace(key, std::make_pair(size_, size));
+        size_ += size;
+    }
+
+    void Commit() noexcept {
+        RuntimeAssert(storage_ == nullptr, "Cannot commit storage twice");
+        storage_ = reinterpret_cast<KRef*>(konanAllocMemory(size_ * sizeof(KRef)));
+    }
+
+    void Clear() noexcept {
+        RuntimeAssert(storage_ != nullptr, "Storage must be committed");
+        for (int i = 0; i < size_; ++i) {
+            UpdateHeapRef(storage_ + i, nullptr);
+        }
+        konanFreeMemory(storage_);
+        map_->clear();
+    }
+
+    KRef* Lookup(Key key, int index) noexcept {
+        RuntimeAssert(storage_ != nullptr, "Storage must be committed");
+        // In many cases there is only one module, so this is one element cache.
+        if (lastKey_ == key) {
+            return storage_ + lastOffset_ + index;
+        }
+        auto it = map_->find(key);
+        RuntimeAssert(it != map_->end(), "Must be there");
+        int offset = it->second.first;
+        RuntimeAssert(offset + index < size_, "Out of bound in TLS access");
+        lastKey_ = key;
+        lastOffset_ = offset;
+        return storage_ + offset + index;
+    }
+
+private:
+    using Map = KStdUnorderedMap<Key, std::pair<int, int>>;
+
+    Map* map_ = nullptr;
```

Comment:
Redudant indirection here?

Reply:
On the other hand, this is not in fact different from the previous version, so should be ok.